File: backend/app/schemas/analytics.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class DailyClicksPoint(BaseModel):
    date: str
    count: int


class TopLinkStat(BaseModel):
    short_code: str
    target_url: str
    clicks: int


class TopCountryStat(BaseModel):
    country: str
    clicks: int


class RecentClickItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    short_code: str
    clicked_at: datetime
    country: str | None
    city: str | None
    device_type: str | None
    browser: str | None
    os: str | None


class AnalyticsOverview(BaseModel):
    clicks_by_day: list[DailyClicksPoint]
    top_link: TopLinkStat | None
    top_country: TopCountryStat | None
    recent_clicks: list[RecentClickItem]

File: frontend/src/types/analytics.ts
export interface DailyClicksPoint {
  date: string;
  count: number;
}

export interface TopLinkStat {
  short_code: string;
  target_url: string;
  clicks: number;
}

export interface TopCountryStat {
  country: string;
  clicks: number;
}

export interface RecentClickItem {
  id: number;
  short_code: string;
  clicked_at: string;
  country: string | null;
  city: string | null;
  device_type: string | null;
  browser: string | null;
  os: string | null;
}

export interface AnalyticsOverview {
  clicks_by_day: DailyClicksPoint[];
  top_link: TopLinkStat | null;
  top_country: TopCountryStat | null;
  recent_clicks: RecentClickItem[];
}
